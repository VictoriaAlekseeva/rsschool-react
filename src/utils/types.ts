import { schema } from './validation';
import * as yup from 'yup';

export interface IFormInput extends yup.InferType<typeof schema> { }
// export type IFormInput = yup.InferType<typeof schema>;
