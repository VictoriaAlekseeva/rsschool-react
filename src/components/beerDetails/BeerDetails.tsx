import React from 'react';
import Image from 'next/image';
import { BeerData } from '../../types/interface';
import Link from 'next/link';
import { useRouter } from 'next/router';

type Props = {
  beerInfo: BeerData[];
};

const BeerDetails: React.FC<Props> = ({ beerInfo }) => {
  const router = useRouter();

  const beer_name = router.query.beer_name || '';
  const page = router.query.page || 1;
  const per_page = router.query.per_page || 5;

  return (
    <div className="beer-card-wrapper">
      <Link href={`/?beer_name=${beer_name}&page=${page}&per_page=${per_page}`}>
        <button> Back </button>
      </Link>
      {beerInfo!.map((item: BeerData) => (
        <div className="beer-card" key={item.id}>
          <h3>{item.name}</h3>
          <div className="beer-card__attr">
            <Image
              width={60}
              height={150}
              className="beer-card__img"
              src={item.image_url}
              alt={item.name}
            />
            <div>
              <ul>
                <li>ABV: {item.abv}%</li>
                <li>IBU: {item.ibu}</li>
                <li>SRM: {item.srm}</li>
              </ul>
              <ul>
                <span className="beer-card__food">Food Pairings:</span>
                {item.food_pairing.map((food) => (
                  <li key={food}>{food}</li>
                ))}
              </ul>
            </div>
          </div>
          <p>{item.description}</p>
        </div>
      ))}
    </div>
  );
};

export default BeerDetails;
