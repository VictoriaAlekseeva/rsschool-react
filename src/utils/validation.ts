import * as yup from 'yup';

export const schema = yup.object().shape({
  name: yup
    .string()
    .matches(/^[A-Z][a-z]*$/, {
      message: 'Name must start with a capital letter',
      excludeEmptyString: true,
    })
    .required('Name is a required field'),
  age: yup
    .number()
    .transform((value) => (isNaN(value) ? undefined : value))
    .nullable()
    .positive('Age must be a positive')
    .integer('Age must be an integer')
    .min(13, 'Must be at least 13 years old')
    .max(120, 'Are you really so old?')
    .required('Age is a required field'),
  email: yup
    .string()
    .email('Invalid email address')
    .required('Email is a required field'),
  password: yup
    .string()
    .matches(/^(?=.*\d)(?=.*[A-Z])(?=.*[a-z])(?=.*[@#$%^&+=!])(?=\S+$).{8,}$/, {
      message:
        'Password must contain at least one digit, one uppercase letter, one lowercase letter, and one special character',
      excludeEmptyString: true,
    })
    .required('Password is a required field'),
  checkpassword: yup
    .string()
    .oneOf([yup.ref('password')], 'Passwords must match')
    .required('Password confirmation is required'),
  checkbox: yup
    .boolean()
    .oneOf([true], 'You must accept the Terms & Conditions')
    .required(),
  gender: yup.string().oneOf(['male', 'female', 'other']).default(null),
  picture: yup
    .mixed()
    .test({
      name: 'fileSize',
      message: 'File size is too large',
      test: (value) => {
        if (!value || !(value instanceof FileList)) return true;
        const kb = 1024;
        const maxSizeMb = 2 * kb * kb;
        const fileSize = (value as FileList)[0]?.size;
        return fileSize <= maxSizeMb;
      },
    })
    .test({
      name: 'fileType',
      message: 'Invalid file format. Only png and jpeg are allowed',
      test: (value) => {
        if (!value || !(value instanceof FileList)) return true;
        const allowedExtensions = ['image/png', 'image/jpeg'];
        const fileType = (value as FileList)[0]?.type;
        return allowedExtensions.includes(fileType);
      },
    }),
  country: yup.string().required('Country is a required field'),
});
