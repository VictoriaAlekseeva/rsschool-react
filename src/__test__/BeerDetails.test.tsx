import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { RouterContext } from 'next/dist/shared/lib/router-context.shared-runtime';

import { BeerData } from '@/__test__/utils/mockData';
import createMockRouter from '@/__test__/utils/mockRouter';
import BeerDetails from '@/components/beerDetails/BeerDetails';

describe('Tests for the Detailed Card component', () => {
  const id = '4';
  const mockRouter = createMockRouter({
    pathname: `/detailedpage/${id}`,
    query: { beer_name: '', page: '1', per_page: '10' },
  });

  it('Make sure the detailed card component correctly displays the detailed card data', async () => {
    const { asFragment } = render(
      <RouterContext.Provider value={mockRouter}>
        <BeerDetails beerInfo={[BeerData[0]]} />
      </RouterContext.Provider>
    );
    const buttonBack = screen.getByText(/Back/i);
    expect(asFragment()).toMatchSnapshot();
    await waitFor(() => {
      expect(buttonBack).toBeInTheDocument();
    });
  });

  it('Ensure that clicking the close button hides the component', async () => {
    render(
      <RouterContext.Provider value={mockRouter}>
        <BeerDetails beerInfo={[BeerData[0]]} />
      </RouterContext.Provider>
    );

    const buttonBack = screen.getByText(/Back/i);

    fireEvent.click(buttonBack);

    await waitFor(() => {
      expect(mockRouter.push).toHaveBeenCalled();
    });
  });
});
