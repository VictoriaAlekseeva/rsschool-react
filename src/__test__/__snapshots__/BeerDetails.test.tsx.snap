// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Component displays detailed card data correctly 1`] = `
<DocumentFragment>
  <div
    class="beer-card-wrapper"
  >
    <a
      href="/?beer_name=&page=1&per_page=10"
    >
      <button>
         Back 
      </button>
    </a>
    <div
      class="beer-card"
    >
      <h3>
        Pilsen Lager
      </h3>
      <div
        class="beer-card__attr"
      >
        <img
          alt="Pilsen Lager"
          class="beer-card__img"
          data-nimg="1"
          decoding="async"
          height="150"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Fimages.punkapi.com%2Fv2%2F4.png&w=128&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fimages.punkapi.com%2Fv2%2F4.png&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Fimages.punkapi.com%2Fv2%2F4.png&w=128&q=75 2x"
          style="color: transparent;"
          width="60"
        />
        <div>
          <ul>
            <li>
              ABV: 6.3%
            </li>
            <li>
              IBU: 55
            </li>
            <li>
              SRM: 15
            </li>
          </ul>
          <ul>
            <span
              class="beer-card__food"
            >
              Food Pairings:
            </span>
            <li>
              Spicy crab cakes
            </li>
            <li>
              Spicy cucumber and carrot Thai salad
            </li>
            <li>
              Sweet filled dumplings
            </li>
          </ul>
        </div>
      </div>
      <p>
        Our Unleash the Yeast series was an epic experiment into the differences in aroma and flavour provided by switching up your yeast. We brewed up a wort with a light caramel note and some toasty biscuit flavour, and hopped it with Amarillo and Centennial for a citrusy bitterness. Everything else is down to the yeast. Pilsner yeast ferments with no fruity esters or spicy phenols, although it can add a hint of butterscotch.
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Make sure the detailed card component correctly displays the detailed card data displays detailed card data correctly 1`] = `
<DocumentFragment>
  <div
    class="beer-card-wrapper"
  >
    <a
      href="/?beer_name=&page=1&per_page=10"
    >
      <button>
         Back 
      </button>
    </a>
    <div
      class="beer-card"
    >
      <h3>
        Pilsen Lager
      </h3>
      <div
        class="beer-card__attr"
      >
        <img
          alt="Pilsen Lager"
          class="beer-card__img"
          data-nimg="1"
          decoding="async"
          height="150"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Fimages.punkapi.com%2Fv2%2F4.png&w=128&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fimages.punkapi.com%2Fv2%2F4.png&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Fimages.punkapi.com%2Fv2%2F4.png&w=128&q=75 2x"
          style="color: transparent;"
          width="60"
        />
        <div>
          <ul>
            <li>
              ABV: 6.3%
            </li>
            <li>
              IBU: 55
            </li>
            <li>
              SRM: 15
            </li>
          </ul>
          <ul>
            <span
              class="beer-card__food"
            >
              Food Pairings:
            </span>
            <li>
              Spicy crab cakes
            </li>
            <li>
              Spicy cucumber and carrot Thai salad
            </li>
            <li>
              Sweet filled dumplings
            </li>
          </ul>
        </div>
      </div>
      <p>
        Our Unleash the Yeast series was an epic experiment into the differences in aroma and flavour provided by switching up your yeast. We brewed up a wort with a light caramel note and some toasty biscuit flavour, and hopped it with Amarillo and Centennial for a citrusy bitterness. Everything else is down to the yeast. Pilsner yeast ferments with no fruity esters or spicy phenols, although it can add a hint of butterscotch.
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tests for the Detailed Card component Make sure the detailed card component correctly displays the detailed card data 1`] = `
<DocumentFragment>
  <div
    class="beer-card-wrapper"
  >
    <a
      href="/?beer_name=&page=1&per_page=10"
    >
      <button>
         Back 
      </button>
    </a>
    <div
      class="beer-card"
    >
      <h3>
        Pilsen Lager
      </h3>
      <div
        class="beer-card__attr"
      >
        <img
          alt="Pilsen Lager"
          class="beer-card__img"
          data-nimg="1"
          decoding="async"
          height="150"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Fimages.punkapi.com%2Fv2%2F4.png&w=128&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fimages.punkapi.com%2Fv2%2F4.png&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Fimages.punkapi.com%2Fv2%2F4.png&w=128&q=75 2x"
          style="color: transparent;"
          width="60"
        />
        <div>
          <ul>
            <li>
              ABV: 6.3%
            </li>
            <li>
              IBU: 55
            </li>
            <li>
              SRM: 15
            </li>
          </ul>
          <ul>
            <span
              class="beer-card__food"
            >
              Food Pairings:
            </span>
            <li>
              Spicy crab cakes
            </li>
            <li>
              Spicy cucumber and carrot Thai salad
            </li>
            <li>
              Sweet filled dumplings
            </li>
          </ul>
        </div>
      </div>
      <p>
        Our Unleash the Yeast series was an epic experiment into the differences in aroma and flavour provided by switching up your yeast. We brewed up a wort with a light caramel note and some toasty biscuit flavour, and hopped it with Amarillo and Centennial for a citrusy bitterness. Everything else is down to the yeast. Pilsner yeast ferments with no fruity esters or spicy phenols, although it can add a hint of butterscotch.
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tests for the Detailed Card component displays detailed card data correctly 1`] = `
<DocumentFragment>
  <div
    class="beer-card-wrapper"
  >
    <a
      href="/?beer_name=&page=1&per_page=10"
    >
      <button>
         Back 
      </button>
    </a>
    <div
      class="beer-card"
    >
      <h3>
        Pilsen Lager
      </h3>
      <div
        class="beer-card__attr"
      >
        <img
          alt="Pilsen Lager"
          class="beer-card__img"
          data-nimg="1"
          decoding="async"
          height="150"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Fimages.punkapi.com%2Fv2%2F4.png&w=128&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fimages.punkapi.com%2Fv2%2F4.png&w=64&q=75 1x, /_next/image?url=https%3A%2F%2Fimages.punkapi.com%2Fv2%2F4.png&w=128&q=75 2x"
          style="color: transparent;"
          width="60"
        />
        <div>
          <ul>
            <li>
              ABV: 6.3%
            </li>
            <li>
              IBU: 55
            </li>
            <li>
              SRM: 15
            </li>
          </ul>
          <ul>
            <span
              class="beer-card__food"
            >
              Food Pairings:
            </span>
            <li>
              Spicy crab cakes
            </li>
            <li>
              Spicy cucumber and carrot Thai salad
            </li>
            <li>
              Sweet filled dumplings
            </li>
          </ul>
        </div>
      </div>
      <p>
        Our Unleash the Yeast series was an epic experiment into the differences in aroma and flavour provided by switching up your yeast. We brewed up a wort with a light caramel note and some toasty biscuit flavour, and hopped it with Amarillo and Centennial for a citrusy bitterness. Everything else is down to the yeast. Pilsner yeast ferments with no fruity esters or spicy phenols, although it can add a hint of butterscotch.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
